cpp/src/arrow/gpu/cuda_memory.h

```diff
@@ -67,6 +68,16 @@ class ARROW_EXPORT CudaBuffer : public Buffer {
   /// \return Status
   Status CopyFromHost(const int64_t position, const void* data, int64_t nbytes);
 
+  /// \brief Copy memory from device to device at position
+  /// \param[in] position start position to copy bytes
+  /// \param[in] data the device data to copy
+  /// \param[in] nbytes number of bytes to copy
+  /// \return Status
+  ///
+  /// \note It is assumed that both source and destination device
+  /// memories have been allocated within the same context.
```

Comment:
Currently, no.

In CudaContext we are using cuMemcpyDtoD for copying data within a device.
To copy data between the two devices, one would need to use [cudaMemcpyPeer](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g88fd1245b2cb10d2d30c74900b7dfb9c) that requires the knowledge of another device.

For copying data between different devices, introduction of new methods is required, e.g.:
```
CudaContext::CudaContextImpl::CopyAnotherDeviceToDevice(void* dst, const void* src, const CudaDevice& src_device, int64_t nbytes)
...
CudaBuffer::CopyFromAnotherDevice(const int64_t position, const void* data, const CudaDevice& data_device, int64_t nbytes)
```
See also [this](https://stackoverflow.com/questions/31628041/how-to-copy-memory-between-different-gpus-in-cuda).

Since I don't have access to multiple GPU devices to test this, I would suggest postponing implementing this feature and create the corresponding JIRA issue. Thoughts?

Reply:
Sounds fine to me